src/DiscoveryAnnouncer.coffee

```diff
@@ -1,83 +1,117 @@
+      announcement.announcementId in @announcedIds
+
+  announce: (announcement, callback) =>
     Utils.promiseRetry(
       @attemptAnnounce.bind(@, announcement)
-      @ANNOUNCE_ATTEMPTS
-      @INITIAL_BACKOFF
+      gAnnouceAttempts
+      gInitialBackoff
     ).nodeify(callback)
 
-  removeAnnouncement:(announcement)->
-    delete @announcements[announcement.announcementId]
+  resetRetryCounts: () ->
+    resetGlobalCounts()
+
+  setAnnouceAttemptsCount: (attemptsCount) ->
+    gAnnouceAttempts = attemptsCount
+
+  setInitialBackoffCount: (backoffCount) ->
+    gInitialBackoff = backoffCount
 
-  attemptAnnounce:(announcement)=>
-    announcement.announcementId or= Utils.generateUUID()
+  removeAnnouncement: (announcement) =>
+    @announcementIndex.removeAnnouncements [announcement.announcementId]
+    @_removeAnnouncedId announcement.announcementId
+
+  attemptAnnounce: (announcement) =>
+    announcement.announcementId or = Utils.generateUUID()
+    announcement.environment = @announceRegion
     @server = @serverList.getRandom()
 
     unless @server
       @reconnect()
       return @discoveryNotifier.notifyAndReject(
-        new Error("Cannot announce. No discovery servers available"))
+        new Error("Cannot announce. No discovery servers available") )
 
     @logger.log "debug", "Announcing " + JSON.stringify(announcement)
     url = @server + "/announcement"
     RequestPromise({
-      url:url
-      method:"POST"
-      json:true
-      body:announcement
-    }).catch(@handleError).then(@handleResponse)
-
+      url: url
+      method: "POST"
+      json: true
+      body: announcement
+    } ).catch(@handleError).then(@handleResponse)
 
-  handleError:(error)=>
+  handleError: (error) =>
     @serverList.dropServer(@server)
     @discoveryNotifier.notifyAndReject(error)
 
-  handleResponse:(response)=>
+  handleResponse: (response) =>
     unless response?.statusCode is 201
       return @discoveryNotifier.notifyAndReject(
-        new Error("During announce, bad status code #{response.statusCode}:#{JSON.stringify(response.body)}"))
+        new Error("During announce, bad status code #{response.statusCode}:#{JSON.stringify(response.body)}") )
     announcement = response.body
-    @logger.log(
-      "info", "Announced as " + JSON.stringify(announcement))
-    @announcements[announcement.announcementId] = announcement
+    @logger.log "info", "Announced as ", JSON.stringify(announcement)
+    @_doAddAnnouncement announcement
     return announcement
 
-  unannounce:(announcement, callback)->
+  unannounce: (announcement, callback) ->
     @attemptUnannounce(announcement)
       .nodeify(callback)
 
-  attemptUnannounce:(announcement)=>
+  _doAddAnnouncement: (announcement) =>
+    @announcementIndex.addAnnouncements [announcement]
+    @_pushAnnouncedId announcement.announcementId
+
+  _pushAnnouncedId: (id) =>
+    if @announcedIds.indexOf(id) < 0
+      @announcedIds.push id
```

Comment:
i must be missing something. doesn't this mean that an announcement is only pushed once?